File: 39_Combination Sum.java
class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private List<Integer> curPath = new ArrayList<>();
    private int[] candidates;

    private void bt(int idx, int left) {
        // end case
        if (left == 0) {
            res.add(new ArrayList<>(curPath));
            return;
        }
        // process
        for (int i = idx; i < candidates.length; i++) {
            // short-cut: terminate if beyond target
            if (candidates[i] > left) break;
            // short-cut: skip to avoid duplicates
            if (i > idx && candidates[i] == candidates[idx]) continue;

            curPath.add(candidates[i]);
            bt(i, left - candidates[i]);
            curPath.remove(curPath.size() - 1);
        }
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        this.candidates = candidates;
        // 1. sort
        Arrays.sort(candidates);
        // 2. process
        bt(0, target);
        return this.res;
    }
}

File: 39_Combination Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        # 完全背包: 每个物品可以用无限次
        # Build a N-nary tree: either not choose or choose from [start:]
        def bt(idx, curSum):
            # end case
            if curSum == target:
                res.append(list(curPath))
                return
            # process
            for i in range(idx, len(candidates)):
                # short-cut: terminate when > target
                if candidates[i] + curSum > target: break
                # short-cut: skip to avoid duplicates
                if i > idx and candidates[i] == candidates[idx]: continue
                curPath.append(candidates[i])
                bt(i, curSum + candidates[i])
                curPath.pop()
        
        # 1. sort
        candidates.sort()
        # 2. bt
        curPath, res = [], []
        bt(0, 0)
        return res